src/Rules/Program.hs

```diff
@@ -49,16 +48,14 @@ buildProgram rs = do
                         -- @llvm-passes@.
                         need =<< ghcDeps stage
 
-                    cross <- flag CrossCompiling
-                    -- For cross compiler, copy @stage0/bin/<pgm>@ to @stage1/bin/@.
-                    case (cross, stage) of
-                        (True, s) | s > Stage0 -> do
-                            srcDir <- buildRoot <&> (-/- (stageString Stage0 -/- "bin"))
-                            copyFile (srcDir -/- takeFileName bin) bin
-                        (False, s) | s > Stage0 && (package `elem` [touchy, unlit]) -> do
-                            srcDir <- stageLibPath Stage0 <&> (-/- "bin")
-                            copyFile (srcDir -/- takeFileName bin) bin
-                        _ -> buildBinary rs bin =<< programContext stage package
+                    -- Packages 'touchy' and 'unlit' are build only in Stage0
```

Comment:
What about the logic for @angerman's beloved cross-compilers? I don't think the first branch of the `case` we had is covered anymore.

Reply:
I think we don't need it any more -- the binaries are now exactly where @angerman wanted them. Or am I wrong?

Reply:
I think we're still missing some logic. For cross-compilers, everything that was under stage0/bin would just be copied if needed for later stages.

Reply:
Hmm, is the exact behaviour documented anywhere? The comment (above) says simply:

```
-- For cross compiler, copy @stage0/bin/<pgm>@ to @stage1/bin/@.
```

Reply:
@angerman Can you give some guidance here?

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
@angerman Thanks! I'm assuming all the above messages are the same, so responding just to the last one :)

> The fundamental reasoning was the following:
> * ghc0 + libs0 + bootstrap0 => libs1, ghc1.
> * ghc1 + libs1 => libs2, ghc2.

I'm trying to translate your notation to what should happen in Hadrian. Is the following correct?

* `ghc0` is what we call Stage0 GHC in Hadrian. It comes with `libs0`.
* `ghc1` is Stage1 GHC, which should be `_build/stage1/bin/ghc` after this PR.
* `ghc2` is Stage2 GHC, which should be `_build/stage2/bin/ghc` after this PR.
* `bootstrap0` live in `_build/stage0/lib` both before and after this PR.
* `libs1` live in `_build/stage1/lib` both before and after this PR.
* `libs2` live in `_build/stage2/lib` both before and after this PR.

Is this correct?

If yes, I'm confused about `libs2`: right now we do not build anything in `_build/stage2/lib`. Do we need to build it?

Reply:
@angerman Thanks! I'm assuming all the above messages are the same, so responding just to the last one :)

> The fundamental reasoning was the following:
> * ghc0 + libs0 + bootstrap0 => libs1, ghc1.
> * ghc1 + libs1 => libs2, ghc2.

I'm trying to translate your notation to what should happen in Hadrian. Is the following correct?

* `ghc0` is what we call Stage0 GHC in Hadrian. It comes with `libs0`.
* `ghc1` is Stage1 GHC, which should be `_build/stage1/bin/ghc` after this PR.
* `ghc2` is Stage2 GHC, which should be `_build/stage2/bin/ghc` after this PR.
* `bootstrap0` live in `_build/stage0/lib` both before and after this PR.
* `libs1` live in `_build/stage1/lib` both before and after this PR.
* `libs2` live in `_build/stage2/lib` both before and after this PR.

Is this correct?

If yes, I'm confused about `libs2`: right now we do not build anything in `_build/stage2/lib`. Do we really need to build it?

Reply:
Let `ghc0` be our bootstrap compiler; and `libs0` the libraries shipped with the bootstrap compiler. `bootstap0` are the libraries we built with `ghc0` to augment `libs0` as we deem them to be potentially too old. Further let `ghcN`, `libsN` be the respective binaries and libraries built in Stage N, by the Stage (N-1) binaries and libraries.

The fundamental reasoning was the following:
- `ghc0` + `libs0` + `bootstrap0` => `libs1`, `ghc1`.
- `ghc1` + `libs1` => `libs2`, `ghc2`.

For cross compilers building `ghc2` or infect any other tool for stage 2 makes no sense, as they would be cross compiled compilers, not cross compilers. That is, for `ghc1` it will run on a different host than the one it most likely targets. `ghc2` build by the same logic would however run on the target and as such not be a cross compiler anymore.

The, we do need `libs2` as those are the libraries against which we will link with `ghc1`; but we *do not* need (or even want) `ghc2` but instead want `ghc1` in it's place.

Does that make sense?

NB: we might want to support the case where you build a cross compiled GHC as well; that is instead of copying `ghc1` into stage2, we would build `ghc2`.

Reply:
lol. Thanks GitHub. 👎

Reply:
@snowleopard yes. that sounds conceptually correct.

I do think we need the stage2 libs (at least for cross compilers; I'm not sure we need stage1 libs?).  If we draw up the dependency graph, by putting the dependencies into parens we'd have:

- `ghc1(ghc0+libs0+bootstrap0)`, this one is `host -> target`, where usually `host == target`.
- `libs1(ghc0+libs0+bootstrap0)`, this one is built for the `host` (by the `ghc1`) compiler.
- `ghc2(ghc1+libs1)`, this one is `target -> target`
- `libs2(ghc1+libs1)`, this one is built for the `target`.

For a cross compiler, we need `libs2` and `ghc1`, as we want to build software for the `target` and thus need the `target` libs to link; we do decidedly *not* need `ghc2`, as we don't want to *run* the compiler on the target.

However, maybe I'm messing this up right now in my head and we never actually build `libs1`? Then `_build/stage1/lib` would be `libs2`.

I feel I'm confused.